### PR TITLE
Fix querying an embedded field in array

### LINKED ISSUE #1724: Querying for an embedded field in array does not behave correctly
### Versions

v0.7.1-29-gaf17fa7-dirty

### What did you do?

See repro script in #1722.

### What did you expect to see?

1 document returned.

### What did you see instead?

0 documents and an assertion.

### Environment

```markdown
- OS: Ubuntu
- Architecture: x86_64
- Version: 22.10
- Go: go1.19.3 linux/amd64
- Git: 2.37.2
- Deployment: Docker
- Deployment details: 20.10.16
```

### DoD

That should be supported by the common Go filtering code. Pushdown for such filters is not in the scope.

- [ ] Add a test data item to `Composites` to support `{a: [{}, {b: 1}]}` example (object -> array -> object)
- [ ] Use `ArrayDocuments`, it has this particular case
- [ ] Run existent test and see if smth needs to be modified because of the new test data item
- [ ] Add compat tests for find, count, findAndModify, update, delete for filter `{'a.b': 1}`
- [ ] If needed, create new issue for problems irrelevant to the current issue
- [ ] Update user docs (dot notation)
- [ ] It may also fix #1828, check it
- [ ] Don't push down queries that have dot notation

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #1724.

## Readiness checklist

* [ ] I added/updated unit tests.
* [x] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [x] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/handlers/common/filter.go [hunk 1]
```diff
@@ -55,85 +55,191 @@ func FilterDocument(doc, filter *types.Document) (bool, error) {
...
 		}
 
-		// we pass the path without the last key because we want {fieldN: *someValue*}, not just *someValue*
-		docValue, err := doc.GetByPath(path.TrimSuffix())
-		if err != nil {
-			return false, nil // no error - the field is just not present
+		if filterSuffix, docs = getDocumentsAtSuffix(doc, path); len(docs) == 0 {
+			// When no document is found at suffix, use an empty one.
+			// So operators such as $nin is applied to the empty document.
+			docs = append(docs, types.MakeDocument(0))
```
Comment (chilagrow) [inline internal/handlers/common/filter.go:70]: Tested by `DocumentDotNotationArrayDocumentNoIndexNin` test case

**Review by rumyantseva — APPROVED** [review]
I don't have a better way to implement it, so I'm ready to approve.

As discussed in the call, maybe we can add some unit tests to showcase the implemented logic (but those unit tests should be excluded from code coverage because we can't trust they have the same logic as MongoDB).

File: internal/handlers/common/filter.go [hunk 2]
```diff
@@ -55,85 +55,191 @@ func FilterDocument(doc, filter *types.Document) (bool, error) {
...
-	}
+	// If we got here, it means that none of the documents matched the filter.
+	return false, nil
+}
 
-	switch filterValue := filterValue.(type) {
-	case *types.Document:
-		// {field: {expr}} or {field: {document}}
-		return filterFieldExpr(doc, filterKey, filterValue)
+// getDocumentsAtSuffix finds documents at the suffix of the given path.
```
Comment (w84thesun) [inline internal/handlers/common/filter.go:144]: Let's add a comment that we go iteratively through the document to find all those matches.

**Review by w84thesun — COMMENTED** [review]
Thanks for all the explanations! I'm pretty happy with the changes, but as discussed small comment for the `getDocumentsAtSuffix` function would help.

**Review by noisersup — DISMISSED** [review]
Thank you for explanation. Great job with implementing this logic.

I see that pg integration tests are failing due to the timeout (same as in other PRs), so it might be necessary to increase it.

File: internal/handlers/common/filter.go (hunk 2 above)
Comment (chilagrow) [reply internal/handlers/common/filter.go]: I added comment and sample example cases in doc string.

**Comment by chilagrow** [comment]
> As discussed in the call, maybe we can add some unit tests to showcase the implemented logic (but those unit tests should be excluded from code coverage because we can't trust they have the same logic as MongoDB).

@rumyantseva  For now, I added show case in the doc string, downside is it might get outdated. I think we maybe able to use `Example....` instead of `Test...` https://go.dev/blog/examples, but not gonna figure out the code coverage related thing in this issue :see_no_evil:

**Review by rumyantseva — APPROVED** [review]
This is one of the most complicated path-related logical parts in FerretDB, I hope the docs will help our future contributors!

**Review by w84thesun — APPROVED** [review]
Thanks for the detailed comment!

**Review by noisersup — APPROVED** [review]
Examples help a lot!

**Pull request merged**